### PR TITLE
test suite

### PR DESCRIPTION
**Pull request by lnhanks**
**What type of PR is this?**

**Which issue does this PR fix**:
None

**What does this PR do / Why do we need it**:
This PR adds a ginkgo test suite that scales the cluster in different ways and sets up the foundation for customers to add their own tests and run them against their cluster.

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this PR introduce any new dependencies?**:

No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
Yes

**Does this change require updates to the CNI daemonset config files to work?**:

No

**Does this PR introduce any user-facing change?**:

No
```release-note
Adds a ginkgo test suite that scales the cluster in different ways and sets up the foundation for customers to add their own tests and run them against their cluster.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/integration/warm-pool/framework.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	. "github.com/onsi/gomega"
+	"time"
+)
+
+var _ = Describe("...", func() {
+	Context("test functions", func() {
+		var testFn func(f *framework.Framework, deployBuilder *manifest.DeploymentBuilder)
+
+		JustBeforeEach(func() {
+			// create deployment and run test case
```
Comment (jdn5126) [inline test/integration/warm-pool/framework.go:17]: nit: Create deployment spec for test function to operate on

File: test/integration/warm-pool/framework.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+				Name("busybox").
+				Namespace(utils.DefaultTestNamespace).
+				Replicas(minPods)
+
+			deployment := deploymentSpec.Build()
+
+			_, err := f.K8sResourceManagers.
+				DeploymentManager().
+				CreateAndWaitTillDeploymentIsReady(deployment, utils.DefaultDeploymentReadyTimeout*5)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (jdn5126) [inline test/integration/warm-pool/framework.go:29]: nit: this can be removed as actually creating the deployment is left to the test case

File: test/integration/warm-pool/framework.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+			start := time.Now().Unix()
+			testFn(f, deploymentSpec)
```
Comment (jdn5126) [inline test/integration/warm-pool/framework.go:32]: We want to get the error and final deployment returned by the test case so that we can check for error and clean up properly. So this should be: `deployment, err := testFn(f, deploymentSpec`, and you will need to change the test function definition to return `(deployment *v1.Deployment, error)`

File: test/integration/warm-pool/framework.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			end := time.Now().Unix()
+
+			createCurlPod()
+			getPrometheusMetrics(start, end)
+			deleteDeployment(deployment)
```
Comment (jdn5126) [inline test/integration/warm-pool/framework.go:37]: Let's add a comment that this deletes the deployment returned by the test case

**Review by jdn5126 — CHANGES_REQUESTED** [review]

File: test/integration/warm-pool/test_fns.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"fmt"
+	"github.com/aws/amazon-vpc-cni-k8s/test/framework"
+	"github.com/aws/amazon-vpc-cni-k8s/test/framework/resources/k8s/manifest"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"strconv"
+)
+
+func QuickScale(f *framework.Framework, deployBuilder *manifest.DeploymentBuilder) {
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:12]: nit: These functions should all return an error and the deployment that they created

File: test/integration/warm-pool/test_fns.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+	Expect(maxPods).To(Equal(busyboxPodCnt()))
+
+	fmt.Fprintf(GinkgoWriter, "Scaling cluster down to %v pods\n", minPods)
+	changeReplicas(deployBuilder, minPods)
+
+	// return deployment so that it can be deleted later
+}
+
+func RandomScale(f *framework.Framework, deployBuilder *manifest.DeploymentBuilder) {
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:25]: nit: let's add comments above each function explaining the test case

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+	minPods = 0               // tests can be run with a base amount of pods at start
+	sleep   = 1 * time.Minute // sleep interval
+)
+
+var clusterIP = "10.100.140.129" // Get the cluster ip of the prometheus-server service
+var primaryInstance *ec2.Instance
+
+var f *framework.Framework
+var err error
+var coreDNSDeploymentCopy *v1.Deployment
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:40]: nit: this can be deleted

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+const CoreDNSDeploymentName = "coredns"
+const KubeSystemNamespace = "kube-system"
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:43]: nit: these can be deleted

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+type Result struct {
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:45]: nit: let's add a comment explaining why we need this struct

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+// Gets the duration in minutes for Prometheus queries
+func strDurationMin(start int64, end int64) string {
+	duration := (end - start) / 60
+	durationMin := strconv.FormatInt(duration, 10)
+	print("TEST DURATION: ", duration)
+	return durationMin
+}
+
+// Random operation, if preventNoChange is 0 this includes no change being a result, otherwise it will add or subtract
+func randOp(replicas int, pods int) (int, string) {
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:171]: nit: let's fix the spacing for this function

**Review by jdn5126 — CHANGES_REQUESTED** [review]

File: config/grafana/dashboard.json [hunk 11]
```diff
@@ -0,0 +1,2404 @@
...
+        }
+      },
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "NdWXluq4z"
+          },
+          "editorMode": "builder",
+          "expr": "delta(awscni_err_no_avail_addrs{k8s_app=\"aws-node\"}[5m])",
```
Comment (jdn5126) [inline config/grafana/dashboard.json:1146]: This needs to be updated after the metrics was renamed to `awscni_no_available_ip_addresses`

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+		fmt.Printf("Cannot unmarshall json: %s", marshallErr)
+	}
+	return result
+}
+
+func getPrometheusMetrics(start int64, end int64) {
+	By("Fetching metrics via Curl Container")
+	// metrics
+	warmMetric := "awscni_total_ip_addresses-awscni_assigned_ip_addresses"
+	noAddrsMetric := "awscni_err_no_avail_addrs"
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:112]: This needs to be updated after the metrics was renamed to `awscni_no_available_ip_addresses`

File: config/grafana/dashboard.json [hunk 13]
```diff
@@ -0,0 +1,2404 @@
...
+        "textMode": "auto"
+      },
+      "pluginVersion": "9.1.0",
+      "targets": [
+        {
+          "datasource": {
+            "uid": "$datasource"
+          },
+          "editorMode": "code",
+          "expr": "count(1 > awscni_eni_util{k8s_app=\"aws-node\"})",
```
Comment (jdn5126) [inline config/grafana/dashboard.json:1898]: This needs to be updated after the metric was renamed to `awscni_assigned_ip_per_eni`

File: config/grafana/dashboard.json [hunk 14]
```diff
@@ -0,0 +1,2404 @@
...
+      },
+      "pluginVersion": "9.1.0",
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "NdWXluq4z"
+          },
+          "editorMode": "builder",
+          "expr": "awscni_eni_util",
```
Comment (jdn5126) [inline config/grafana/dashboard.json:2165]: This needs to be updated after the metric was renamed to `awscni_assigned_ip_per_eni`

File: test/integration/warm-pool/README.md [hunk 15]
```diff
@@ -0,0 +1,12 @@
+## amazon-vpc-cni-k8s Warm Pool Test Suite
+The test framework consists of integration tests using the Ginkgo framework invoked manually. It provides a foundation for evaluation IP consumption.
```
Comment (jdn5126) [inline test/integration/warm-pool/README.md:2]: nit: "The test" -> "This test"
"evaluation" -> "evaluating"

File: test/integration/warm-pool/test_fns.go [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+// Scales the cluster from minPods up to maxPods and back down to minPods
+func QuickScale(f *framework.Framework, deploymentSpec *manifest.DeploymentBuilder) (*v1.Deployment, error) {
+	// create deployment
+	deployment := deploymentSpec.Build()
+
+	_, err := f.K8sResourceManagers.
+		DeploymentManager().
+		CreateAndWaitTillDeploymentIsReady(deployment, utils.DefaultDeploymentReadyTimeout*5)
+	Expect(err).ToNot(HaveOccurred())
+
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:23]: We can skip this part and jump right into scaling up

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+		randPods = rand.Intn(randDigits)
+		result, op = randOp(replicas, randPods)
+		if result > minPods && result < maxPods && randPods != 0 {
+			return result, op, randPods
+		}
+	}
+	return result, op, randPods
+}
+
+func changeReplicas(deploymentSpec *manifest.DeploymentBuilder, pods int) {
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:274]: This function should be returning the `*v1.Deployment` that it creates

File: test/integration/warm-pool/test_fns.go [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+	fmt.Fprintf(GinkgoWriter, "Scaling cluster up to %v pods\n", maxPods)
+	changeReplicas(deploymentSpec, maxPods)
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:25]: nit: this function should return the deployment that it created

File: test/integration/warm-pool/test_fns.go [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+	Expect(maxPods).To(Equal(busyboxPodCnt()))
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:27]: Instead of using `Expect` in these functions, they should return an `error` object so that the caller can catch it

File: test/integration/warm-pool/test_fns.go [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+// or subtract a random amount of pods depending on the value of randDigits. The behavior could also result in no change
+// to the cluster.
+func RandomScale(f *framework.Framework, deploymentSpec *manifest.DeploymentBuilder) (*v1.Deployment, error) {
+	// create deployment
+	deployment := deploymentSpec.Build()
+
+	_, err := f.K8sResourceManagers.
+		DeploymentManager().
+		CreateAndWaitTillDeploymentIsReady(deployment, utils.DefaultDeploymentReadyTimeout*5)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:46]: I think we can delete this and jump right into scaling the pods

File: test/integration/warm-pool/test_fns.go [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+
+	replicas := minPods
+
+	for i := 0; i < iterations; i++ {
+		By("Loop " + strconv.Itoa(i))
+		result, op, randPods := randOpLoop(replicas)
+		replicas = checkInRange(result)
+		if op == "no change" {
+			fmt.Fprintf(GinkgoWriter, "No change to cluster, %v pods", replicas)
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:55]: If there is no change, should we `continue` and skip this iteration?

File: test/integration/warm-pool/test_fns.go [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+// Scales the cluster up proportionately from minPods to maxPods dependent on what the value of scale is set to, and
+// then back down to minPods at the same scale.
+func ProportionateScaling(f *framework.Framework, deploymentSpec *manifest.DeploymentBuilder) (*v1.Deployment, error) {
+	// create deployment
+	deployment := deploymentSpec.Build()
+
+	_, err := f.K8sResourceManagers.
+		DeploymentManager().
+		CreateAndWaitTillDeploymentIsReady(deployment, utils.DefaultDeploymentReadyTimeout*5)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (jdn5126) [inline test/integration/warm-pool/test_fns.go:74]: I think we can skip this part and jump right into the scaling

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+// Warm Pool Test Suite Constants
+// Run all tests with these constants or change individual tests to get desired outcome
+// Environment variables are used in the tests listed in the (...)
+const (
+	randDigits      = 10   // exclusive, used in rand.Intn to change scale amount, <= maxPods, (3,6,9)
+	scale           = 0.25 // used in set proportional scaling, iterate with a fixed percentage (5)
+	iterations      = 2    // run test over a set number of iterations (2,3,4,7,8)
+	iterPods        = 1    // iterate with a fixed number of pods (2,7,8)
+	numBursts       = 2    // Use Case Test 8, set number of bursts (8)
+	preventNoChange = 1    // retries x amount of times if randInt/randOp is out of range, if out of range no cluster
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:29]: This is a double-negative, and I am not sure I follow it. Can we rename this?

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+
+	curlPod := manifest.NewDefaultPodBuilder().
+		Name("curl-pod").
+		Namespace(utils.DefaultTestNamespace).
+		HostNetwork(true).
+		Container(curlContainer).
+		Build()
+
+	testPod, _ := f.K8sResourceManagers.PodManager().
+		CreateAndWaitTillPodCompleted(curlPod)
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:235]: How long does this take? I am a bit worries this is waiting all 1000 seconds for the curl container to complete

File: test/integration/warm-pool/warm_pool_suite_test.go [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+
+	logs, errLogs := f.K8sResourceManagers.PodManager().
+		PodLogs(testPod.Namespace, testPod.Name)
+	Expect(errLogs).ToNot(HaveOccurred())
+	fmt.Fprintln(GinkgoWriter, logs)
+}
+
+func deleteDeployment(deploymentSpec *v1.Deployment) {
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_suite_test.go:243]: nit: "deploymentSpec" -> "deployment"

File: test/integration/warm-pool/warm_pool_test.go [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+				Name("busybox").
+				Namespace(utils.DefaultTestNamespace).
+				Replicas(minPods)
+
+			start := time.Now().Unix()
+			// run test function
+			deployment, err := testFn(f, deploymentSpec)
+			if err != nil {
+
+			}
```
Comment (jdn5126) [inline test/integration/warm-pool/warm_pool_test.go:31]: You can change this to an `Expect(err).ToNot(HaveOccurred())`

**Review by jdn5126 — CHANGES_REQUESTED** [review]
This is starting to come together, thanks!

File: test/integration/warm-pool/test_fns.go (hunk 16 above)
Comment (lnhanks) [reply test/integration/warm-pool/test_fns.go:23]: Screenshot 2023-09-15 at 9 28 23 AM
If I'm understanding this correctly, you mean to skip the CreateAndWait?

Comment (jdn5126) [reply test/integration/warm-pool/test_fns.go:23]: Skip that and the `deploymentSpec.Build`, as `changeReplicas` should be the only one to build the deployment

Comment (lnhanks) [reply test/integration/warm-pool/test_fns.go:23]: Okay, so you want changeReplicas to create the deployment and then update the deployment as needed? I think I'm just getting thrown up incorporating the two different functions into changeReplicas.

File: test/integration/warm-pool/warm_pool_suite_test.go (hunk 24 above)
Comment (lnhanks) [reply test/integration/warm-pool/warm_pool_suite_test.go:235]: I think I might have been understanding this incorrectly, but I was following existing code in pod_egress_test to create a curl pod and use it to run multiple commands later. However, I am creating the curl pod after the test is completed and not before to not impact the test. I was under the impression that the pod would sleep and be available for any commands in x amount of time, but as running the Prometheus queries doesn't take long the curl pod will be deleted when the command to delete the curl pod runs.  If I don't include some kind of a sleep, I get an error "unable to upgrade connection: container not found." But I think I may be misunderstanding how this works and there is a better solution here. 

tldr; It's never taken 1000 seconds, I do have a time.sleep with changeReplicas mostly to be able to better visualize metrics in the dashboard.

Comment (lnhanks) [reply test/integration/warm-pool/warm_pool_suite_test.go:243]: All other comments should be resolved but this this one.

**Comment by jdn5126** [comment]
Closing this pull request for now as we are considering a different direction for capturing these metrics.
